### ISSUE TITLE
[Enhancement]: Extending Testcontainers to support multiple container runtimes (Podman)

### ISSUE THREAD

**Issue by HofmeisterAn** [testcontainers/testcontainers-dotnet#876]
### Problem

Since Testcontainers for Node has done an awesome job supporting [multiple container runtimes](https://github.com/testcontainers/testcontainers-node/actions/runs/4763812640) (not just Docker), it makes perfect sense to continue this great work and look into implementing it for .NET. We can start by collecting a list of tasks, issues or incompabilities that are necessary to support other runtimes as well.

- [x] `Docker.DotNet`'s `ExtractArchiveToContainerAsync(string, ContainerPathStatParameters, Stream, CancellationToken)` throws `System.Net.Sockets.SocketException : Connection reset by peer`.
    - ✅ https://github.com/dotnet/Docker.DotNet/pull/629
- [x] Running Testcontainers for .NET on **macOS** using Podman (brew installation) does not work well with the Resource Reaper. Testcontainers is incapable of mounting the daemon socket. It fails with different errors:
    ```
    Docker API responded with status code=InternalServerError, response={"cause":"operation not supported","message":"container create: statfs /Users/atomicjar/.local/share/containers/podman/machine/qemu/podman.sock: operation not supported","response":500}
    ```
    When attempting to use the daemon socket path from within the Podman virtual machine, the process is also unsuccessful. Setting `export TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/var/run/docker.sock` fails:
    ```
    Docker API responded with status code=InternalServerError, response={"cause":"permission denied","message":"container create: statfs /var/run/docker.sock: permission denied","response":500}
    ```
    When running the Podman virtual machine with rootful privileges (`podman machine set --rootful=true`), the errors mentioned above do not occur. However, Testcontainers for .NET is either unable to establish a connection to the Resource Reaper or fails:
    ```
    Docker API responded with status code=NotFound, response={"cause":"no such container","message":"no container with name or ID \"b26ef6bf452c4d85051d7727a2f53b2dddcf0c55b824aa03a6dd20b203ebc6f1\" found: no such container","response":404}
    ```
    - ✅ To fix the issue, you need to run the Podman virtual machine with root privileges and set the environment variables `TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/var/run/docker.sock` and `TESTCONTAINERS_RYUK_CONTAINER_PRIVILEGED=true`.
- [ ] `K3sContainer` does not start (`rancher/k3s:v1.26.2-k3s1`)
    ```console
    Failed to load kernel module br_netfilter with modprobe
    Failed to load kernel module iptable_nat with modprobe
    Failed to load kernel module iptable_filter with modprobe
    Failed to set sysctl: open /proc/sys/net/bridge/bridge-nf-call-iptables: no such file or directory
    Failed to set sysctl: open /proc/sys/net/netfilter/nf_conntrack_max: permission denied
    Failed to ApplyOOMScoreAdj" err="write /proc/self/oom_score_adj: permission denied
    Failed to set rlimit on max file handles" err="operation not permitted
    Failed to get the info of the filesystem with mountpoint" err="unable to find data in memory cache" mountpoint="/var/lib/rancher/k3s/agent/containerd/io.containerd.snapshotter.v1.overlayfs
    Failed to start ContainerManager" err="[open /proc/sys/vm/overcommit_memory: permission denied, open /proc/sys/kernel/panic: permission denied, open /proc/sys/kernel/panic_on_oops: permission denied]
    [...]
    ```
- [ ] `LocalStackContainer` does not start (`localstack/localstack:1.4`, `localstack/localstack:2.0`)
    ```
    2023-05-17T14:17:05.033  WARN --- [   asgi_gw_0] localstack.utils.archives  : Attempt 1. Failed to download archive from https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip: MyHTTPSConnectionPool(host='s3-us-west-2.amazonaws.com', port=443): Max retries exceeded with url: /dynamodb-local/dynamodb_local_latest.zip (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0xffff65444bb0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
    2023-05-17T14:17:05.037  INFO --- [   asgi_gw_0] localstack.utils.archives  : Unable to extract file, re-downloading ZIP archive /tmp/localstack.ddb.zip: ('Failed to download archive from %s: . Retries exhausted', 'https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip')
    2023-05-17T14:17:25.079  WARN --- [   asgi_gw_0] localstack.utils.archives  : Attempt 1. Failed to download archive from https://s3-us-west-2.amazonaws.com/dynamodb-local/dynamodb_local_latest.zip: MyHTTPSConnectionPool(host='s3-us-west-2.amazonaws.com', port=443): Max retries exceeded with url: /dynamodb-local/dynamodb_local_latest.zip (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0xffff65444d00>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
    2023-05-17T14:17:25.080  WARN --- [   asgi_gw_0] localstack.utils.functions : error calling function on_before_start: Installation of dynamodb-local failed.
    [...]
    ```
- [ ] Failed test:
    ```console
      Failed DotNet.Testcontainers.ResourceReaper.Tests.DefaultResourceReaperTest.ContainerCleanUpStartsDefaultResourceReaper(resourceReaperEnabled: True) [1 ms]
      Error Message:
      Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"cause":"no such container","message":"container 02bb9d49a7b7ecd0501c4e1035c544c5ba77cd90b5596415a06be7dfc0cdf50a does not exist in database: no such container","response":500}

      Stack Trace:
        at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response, IEnumerable`1 handlers) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 447
      at Docker.DotNet.DockerClient.MakeRequestAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 217
      at Docker.DotNet.ContainerOperations.InspectContainerAsync(String id, CancellationToken cancellationToken) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/Endpoints/ContainerOperations.cs:line 76
      at DotNet.Testcontainers.Containers.DockerContainer.UnsafeStopAsync(CancellationToken ct) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 442
      at DotNet.Testcontainers.Containers.DockerContainer.DisposeAsyncCore() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 311
      at DotNet.Testcontainers.Resource.DisposeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Resource.cs:line 27
      at DotNet.Testcontainers.Containers.ResourceReaper.DisposeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/ResourceReaper.cs:line 170
      at DotNet.Testcontainers.ResourceReaper.Tests.DefaultResourceReaperTest.InitializeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.ResourceReaper.Tests/DefaultResourceReaperTest.cs:line 17
    ```
- [ ] Failed test:
    ```console
      Failed DotNet.Testcontainers.ResourceReaper.Tests.DefaultResourceReaperTest.ContainerCleanUpStartsDefaultResourceReaper(resourceReaperEnabled: False) [1 ms]
      Error Message:
      Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"cause":"no such container","message":"container deca45c9a28733786162a3777bf145d0d00d37f335f7597d115a02186d5dd2a2 does not exist in database: no such container","response":500}

      Stack Trace:
        at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response, IEnumerable`1 handlers) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 447
      at Docker.DotNet.DockerClient.MakeRequestAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 217
      at Docker.DotNet.ContainerOperations.InspectContainerAsync(String id, CancellationToken cancellationToken) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/Endpoints/ContainerOperations.cs:line 76
      at DotNet.Testcontainers.Containers.DockerContainer.UnsafeStopAsync(CancellationToken ct) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 442
      at DotNet.Testcontainers.Containers.DockerContainer.DisposeAsyncCore() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 311
      at DotNet.Testcontainers.Resource.DisposeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Resource.cs:line 27
      at DotNet.Testcontainers.Containers.ResourceReaper.DisposeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/ResourceReaper.cs:line 170
      at DotNet.Testcontainers.ResourceReaper.Tests.DefaultResourceReaperTest.InitializeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.ResourceReaper.Tests/DefaultResourceReaperTest.cs:line 17
    ```
- [x] The test runs either with Docker or Podman if the `mtu` instead of the `com.docker.network.driver.mtu` key is used. Is the `mtu` a valid key for Docker (it is probably not that important since we only test if TC sets the property)?
    ```console
      Failed DotNet.Testcontainers.Tests.Unit.TestcontainerNetworkBuilderTest.CreateNetworkAssignsOptions [38 ms]
      Error Message:
      Assert.Contains() Failure
    Not found: com.docker.network.driver.mtu
    In value:  KeyCollection<String, String> ["mtu"]
      Stack Trace:
        at DotNet.Testcontainers.Tests.Unit.TestcontainerNetworkBuilderTest.CreateNetworkAssignsOptions() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Networks/TestcontainerNetworkBuilderTest.cs:line 59
    ```
- [x] Cannot ping containers by its hostname (`--hostname`, `WithHostname(string)`), the ping commands fails with `ping: bad address '_container2'`.
    https://github.com/testcontainers/testcontainers-dotnet/blob/f0fef3ee3346665842426f1fe4bdead322c0097e/tests/Testcontainers.Tests/Unit/Networks/TestcontainersNetworkTest.cs#L47
    ```console
      Failed DotNet.Testcontainers.Tests.Unit.TestcontainersNetworkTest.PingContainer(destination: "_testcontainer2") [85 ms]
      Error Message:
      Assert.Equal() Failure
    Expected: 0
    Actual:   1
      Stack Trace:
        at DotNet.Testcontainers.Tests.Unit.TestcontainersNetworkTest.PingContainer(String destination) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Networks/TestcontainersNetworkTest.cs:line 52
    ```
- [x] Failed test:
    - This looks like a race condition after the tests are finished (disposing the container).
    ```console
      Failed DotNet.Testcontainers.Tests.Unit.TestcontainersContainerCancellationTest+Cancel.Start [1 ms]
      Error Message:
      [Test Class Cleanup Failure (DotNet.Testcontainers.Tests.Unit.TestcontainersContainerCancellationTest+Cancel)]: Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"cause":"no such container","message":"container 4c86760f485d02cb2fe7d1a20527fdc7c81fe6a9d24be63a3b16ea4884085310 does not exist in database: no such container","response":500}

      Stack Trace:
        at Docker.DotNet.DockerClient.HandleIfErrorResponseAsync(HttpStatusCode statusCode, HttpResponseMessage response, IEnumerable`1 handlers) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 447
      at Docker.DotNet.DockerClient.MakeRequestAsync(IEnumerable`1 errorHandlers, HttpMethod method, String path, IQueryString queryString, IRequestContent body, IDictionary`2 headers, TimeSpan timeout, CancellationToken token) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/DockerClient.cs:line 217
      at Docker.DotNet.ContainerOperations.InspectContainerAsync(String id, CancellationToken cancellationToken) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Docker.DotNet/Endpoints/ContainerOperations.cs:line 76
      at DotNet.Testcontainers.Containers.DockerContainer.UnsafeStopAsync(CancellationToken ct) in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 442
      at DotNet.Testcontainers.Containers.DockerContainer.DisposeAsyncCore() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Containers/DockerContainer.cs:line 311
      at DotNet.Testcontainers.Resource.DisposeAsync() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/src/Testcontainers/Resource.cs:line 27
    ```
- [ ] Podman can only set the mac address if the container is [assigned to a network](https://docs.podman.io/en/latest/markdown/options/mac-address.html). `.DependsOn(new NetworkBuilder().Build())` works. The Podman "workaround" does not work with Docker.
    > This option can only be used if the <<container|pod>> is joined to only a single network
    ```console
      Failed DotNet.Testcontainers.Tests.Unit.TestcontainersContainerTest+WithConfiguration.MacAddress [899 ms]
      Error Message:
      Assert.Equal() Failure
              ↓ (pos 0)
    Expected: 92:95:5e:30:fe:6d
    Actual:   ea:2c:b1:6c:16:c5
              ↑ (pos 0)
      Stack Trace:
        at DotNet.Testcontainers.Tests.Unit.TestcontainersContainerTest.WithConfiguration.MacAddress() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Containers/Unix/TestcontainersContainerTest.cs:line 107
      at DotNet.Testcontainers.Tests.Unit.TestcontainersContainerTest.WithConfiguration.MacAddress() in /home/runner/work/testcontainers-dotnet/testcontainers-dotnet/tests/Testcontainers.Tests/Unit/Containers/Unix/TestcontainersContainerTest.cs:line 108
    ```

### Solution

\-

### Benefit

\-

### Alternatives

\-

### Would you like to help contributing this enhancement?

Yes

**Comment by tebeco** [comment]
Context:
Apple MacBook Pro M3 - ARM
Podman installed

might be a pre-requisite:
```console
brew update
brew upgrade

# this is because QEMU had a bug since mid december / january that pulled an x86 image on ARM
# this is likely fixed and force a full re-creation
podman machine stop
podman machine rm
podman machine init

podman machine start
```

* did not touch any env var
* did not had to create any `ssh` tunnel

All I had to do was:
```console
$ echo "ryuk.container.privileged = true" >> $HOME/.testcontainers.properties
$ cat $HOME/.testcontainers.properties

ryuk.container.privileged = true
$
```
and from `podman` Dashboard enable `Docker compatibility`:
<img width="1004" alt="image" src="https://github.com/testcontainers/testcontainers-dotnet/assets/2266487/97c6de41-087f-4e76-8209-b947abe43e8a">

**Comment by kanpov** [comment]
I've tried using Testcontainers with Podman via a simple `ln /run/user/1000/podman/podman.sock /run/user/1000/docker.sock` and so far it works exactly as well as with Docker, so are the issues only related to tests or are there any functional hindrances?

**Comment by HofmeisterAn** [comment]
> so are the issues only related to tests or are there any functional hindrances?

The issues are not just related to the tests in this repository. Depending on your Podman environment and configuration, they can happen their as well.

**Comment by tebeco** [comment]
should this still be opened @HofmeisterAn ?

since qemu fixed their image

**Comment by HofmeisterAn** [comment]
I'm not sure, but in my opinion, we should keep the issue open until we have a working GitHub workflow that supports Podman. I haven't recently tested TC for .NET with Podman, I don't know which features are working and which are not.

**Comment by tebeco** [comment]
i fail to understand the relation between "testcontainer works with podman"✅ versus "adding a github workflow integration feature (to testcontainer withpod man")⚒️

that second one seems like an integration evol one that needs to be created as "proposal"
while the second one seems working (this ticket)

**Comment by HofmeisterAn** [comment]
I think there are still a few issues that prevent Testcontainers for .NET from working properly with Podman, or at least obstacles. Ideally, this issue addresses the Podman incompatibilities (failing tests) and establishes a proper GH workflow (incl. updating the docs and explain which steps are necessary to set it up). Otherwise, we cannot know which features are working, which are not, and which may break with newer versions.

**Comment by tebeco** [comment]
/remindme 1year

🙃 (bad joke ok ...)

**Comment by mdx0111** [comment]
To get it working on Fedora 40, I ran the following command
`echo "ryuk.container.privileged = true" >> $HOME/.testcontainers.properties`

Then ran this
`ln /run/user/1000/podman/podman.sock /run/user/1000/docker.sock`

And everything started working nicely.

**Comment by serezlan** [comment]
Hi,

I'm facing problem when I create network with TC.

Steps too reproduce:
- create INetwork 

Result:
   Docker.DotNet.DockerApiException : Docker API responded with status code=InternalServerError, response={"cause":"CNI network \"1ddde97e-1272-4bf4-82c3-4998f4ab8c08\" not found","message":"error configuring network namespace for container aaeba79b34ab9de814ea4fa586696d507d72d47fd566c147748321f7a1285873: CNI network \"1ddde97e-1272-4bf4-82c3-4998f4ab8c08\" not found","response":500}

Environment:
TC version : 3.10.0
dotnet 8
debian 12

I'm not sure if this is podman or TC issue. Please let me know if I need to make changes.

**Comment by serezlan** [comment]
apparently the default podman version in my distro is quite old (v3)
I upgrade to v4.3.1 and I got it working.

thx :)

**Comment by HofmeisterAn** [comment]
I would like to share the following article for anyone experiencing issues with setting up Podman together with Testcontainers: https://podman-desktop.io/tutorial/testcontainers-with-podman.

**Comment by HofmeisterAn** [comment]
During my lunch break, I had some time and decided to try out Podman. Installing and running Podman Desktop on Windows was pretty smooth. After installing Podman Desktop, I skipped all the initial setup steps, installed Podman itself, and initialized the Podman Machine (WSL). Testcontainers immediately detected the container runtime and was able to start Ryuk and an Alpine container without any issues (Docker Desktop is installed as well, but it was not running).

**Comment by haiddablitz** [comment]
I have also just tested Podman. The Ryuk image is downloaded, as well as the desired image of the test container. As WaitingStrategy I use `Wait.ForUnixContainer().UntilContainerIsHealthy()`.

However, the tests abort immediately after starting with a `System.TimeoutException`.

I suspect a problem with the Docker compatibility mode and the health checks, but I don't currently know how to verify this assumption.

**Comment by haiddablitz** [comment]
Just played around with Podman again. After switching from `.UntilContainerIsHealthy()` to `.UntilPortIsReady()` everything works fine.

**Comment by vlflorian** [comment]
@haiddablitz We are having the same issue with Podman; `System.TimeoutException `after 1ms.             `.WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(6379))`
 unfortunately did not help. Did you have to set anything else up to make it work? 
Thanks!

**Comment by HofmeisterAn** [comment]
Which timeout exception do you get? Please share the stack trace (it's a bit off-topic though)? Just a heads-up. The [wait strategy](https://github.com/testcontainers/testcontainers-dotnet/blob/2162c445ab6624b42d9d7945b0652da5bc56663c/src/Testcontainers/Configurations/WaitStrategies/UntilUnixPortIsAvailable.cs#L6) relies on certain Unix commands, it won't work with a scratch image.

**Comment by vlflorian** [comment]
@HofmeisterAn You're right, it is maybe a bit off-topic, my apologies. I pasted the stack trace [here](https://github.com/testcontainers/testcontainers-dotnet/issues/1438) in a separate issue.

**Comment by cremor** [comment]
Microsoft announced that WSL will get its own container runtime: [WSL container](https://learn.microsoft.com/en-us/windows/wsl/wsl-container)
I hope Testcontainers will work with that.

If for some reason the new runtime isn't compatible with the Docker API then the alternative would be to use an official (not yet released) NuGet Package which will provide an API. But I understand that that would be a big change for Testcontainers.

**Comment by HofmeisterAn** [comment]
@cremor Thanks for the heads-up. I've opened an issue in Docker.DotNet so we can evaluate whether we can integrate it: https://github.com/testcontainers/Docker.DotNet/issues/113.

**Comment by glen-84** [comment]
Adding a concrete ask for the rootless-Podman side of this: **first-class unprivileged reaping** — and why it matters beyond cleanup.

On rootless Podman, Ryuk currently needs `TESTCONTAINERS_RYUK_PRIVILEGED=true` (socket access under user namespaces, SELinux on some hosts, the `bridge` vs `podman` default-network naming, etc.), which undercuts the whole reason for running rootless in the first place. The usual escape hatch is `TESTCONTAINERS_RYUK_DISABLED=true`, and for ordinary cleanup that's fine — `DisposeAsync` removes the container in the normal case.

The catch: disabling Ryuk silently takes `WithReuse` off the table. With no reaper session `SessionId` is `Guid.Empty`, so `DockerContainer.DisposeAsyncCore` takes the stop branch regardless of `Reuse`:

```csharp
if (Guid.Empty.Equals(_configuration.SessionId))
    await UnsafeStopAsync();   // stopped on dispose, even when Reuse == true
else
    await UnsafeDeleteAsync();
```

So a reused container is stopped at the end of every run and never survives to be re-attached; the next run attaches to a stopped/recovering container. (For a SQL Server fixture this surfaces as EF `EnsureCreated` seeing a not-yet-openable database and trying to recreate it — `CREATE DATABASE ... already exists`.) In other words, on rootless Podman `WithReuse` and Ryuk-disabled are effectively mutually exclusive today.

So I'd love to see reaping work **unprivileged** on rootless Podman — it would give safe cleanup without `--privileged` *and* unblock `WithReuse` for Podman users in one go, since the two are coupled through the reaper. Happy to help test on rootless Podman 5.x if that's useful.

_Drafted by Claude (Anthropic AI assistant)._
